In [237]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import plotly.express as px

In [151]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [152]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [153]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [154]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [155]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

market_data_scaled= StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y']])

In [162]:
# Create a DataFrame with the scaled data

df_market_data_scaled = pd.DataFrame(market_data_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

# Copy the crypto names from the original data

df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data

df_market_data_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

In [163]:
df_market_data_scaled.info()

<class 'pandas.core.frame.DataFrame'>
Index: 41 entries, bitcoin to digibyte
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   price_change_percentage_24h   41 non-null     float64
 1   price_change_percentage_7d    41 non-null     float64
 2   price_change_percentage_14d   41 non-null     float64
 3   price_change_percentage_30d   41 non-null     float64
 4   price_change_percentage_60d   41 non-null     float64
 5   price_change_percentage_200d  41 non-null     float64
 6   price_change_percentage_1y    41 non-null     float64
dtypes: float64(7)
memory usage: 2.6+ KB


In [164]:
df_market_data_scaled.index

Index(['bitcoin', 'ethereum', 'tether', 'ripple', 'bitcoin-cash',
       'binancecoin', 'chainlink', 'cardano', 'litecoin', 'bitcoin-cash-sv',
       'crypto-com-chain', 'usd-coin', 'eos', 'monero', 'tron', 'tezos', 'okb',
       'stellar', 'cosmos', 'cdai', 'neo', 'wrapped-bitcoin', 'leo-token',
       'huobi-token', 'nem', 'binance-usd', 'iota', 'vechain', 'zcash',
       'theta-token', 'dash', 'ethereum-classic', 'ethlend', 'maker', 'havven',
       'omisego', 'celsius-degree-token', 'ontology', 'ftx-token', 'true-usd',
       'digibyte'],
      dtype='object', name='coin_id')

### Find the Best Value for k Using the Original Data.

In [165]:
# Create a list with the number of k-values from 1 to 11
k=list(range(1,11))

In [166]:
# Create an empty list to store the inertia values

inertia=[]
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


c:\Users\lparedes\Anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [167]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)


In [220]:
df_elbow.head()

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [205]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
kmeans_elbow_plot =df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve - KMeans - Scaled Market Data", 
    xticks=k
)

kmeans_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [206]:
# Initialise the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [207]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

KMeans(n_clusters=4)

In [208]:
# Predict the clusters to group the cryptocurrencies using the scaled data

k_3 = model.predict(df_market_data_scaled)
# Print the resulting array of cluster values.

print(k_3)


[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [172]:
# Create a copy of the DataFrame

df_market_data_scaled_predictions = df_market_data_scaled.copy()

In [209]:
# Add a new column to the DataFrame with the predicted clusters

df_market_data_scaled_predictions["CoinCluster"]= k_3
# Display sample data

df_market_data_scaled_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  CoinCluster  
coin_id                                                
bitcoin                        -0.251637            3  
ethereum                       -0.199352            3  
tether                         -0.282061            0  
ripple                         -0.295546            0  
bitcoin-cash                   -0.270317            3

In [222]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

scatter_scaled_clusters = df_market_data_scaled_predictions.hvplot.scatter(

    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="CoinCluster",
    hover_cols =["coin_id"],
    title = "Scatter Plot by KMeans Scaled Market Data Clusters"
)

scatter_scaled_clusters


:NdOverlay   [CoinCluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

In [242]:
# 3D Plot

df_market_scaled_3d = df_market_data_scaled_predictions.copy()

df_market_scaled_3d["coin_id"] = df_market_data.index



fig_kmeans = px.scatter_3d(
    df_market_scaled_3d,
    hover_name="coin_id",
    hover_data=["CoinCluster"],
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    z="price_change_percentage_14d",
    color="CoinCluster",
    width=800,
)

# fig.update_layout(legend)
fig_kmeans.show()

---

### Optimise Clusters with Principal Component Analysis.

In [211]:
# Create a PCA model instance and set `n_components=3`.

pca = PCA(n_components=3)

In [212]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

market_pca_data = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 

market_pca_data [:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [213]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
var = pca.explained_variance_ratio_

print(var)

sum(var)


[0.3719856  0.34700813 0.17603793]


0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** About 89.5% of the total variance is condensed into the 3 PCA variables.

In [214]:
# Create a new DataFrame with the PCA data.

df_market_pca_data =pd.DataFrame(market_pca_data, columns=["PC1","PC2","PC3"])

# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data

df_market_pca_data["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_pca_data = df_market_pca_data.set_index("coin_id")


# Display sample data

df_market_pca_data.head()


PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [215]:
# Create a list with the number of k-values from 1 to 11

k=list(range(1,11))

In [216]:
# Create an empy list to store the inertia values

inertia=[]
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_pca_data)
    inertia.append(k_model.inertia_)

c:\Users\lparedes\Anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [217]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [218]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

pca_elbow_plot = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve - KMeans - Market PCA Data", 
    xticks=k
)

pca_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [188]:
# Initialise the K-Means model using the best value for k

model = KMeans(n_clusters=4)

In [189]:
# Fit the K-Means model using the PCA data

model.fit(df_market_pca_data)

KMeans(n_clusters=4)

In [190]:
# Predict the clusters to group the cryptocurrencies using the PCA data

crypto_clusters  = model.predict(df_market_pca_data)

# Print the resulting array of cluster values.

print(crypto_clusters)


[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [192]:
# Create a copy of the DataFrame with the PCA data

df_market_predictions_pca_data = df_market_pca_data.copy()

# Add a new column to the DataFrame with the predicted clusters

df_market_predictions_pca_data["CryptoClusters"] = crypto_clusters
# Display sample data

df_market_predictions_pca_data.head()

PC1       PC2       PC3  CryptoClusters
coin_id                                                   
bitcoin      -0.600667  0.842760  0.461595               3
ethereum     -0.458261  0.458466  0.952877               3
tether       -0.433070 -0.168126 -0.641752               0
ripple       -0.471835 -0.222660 -0.479053               0
bitcoin-cash -1.157800  2.041209  1.859715               3

In [225]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


scatter_pca_clusters = df_market_predictions_pca_data.hvplot.scatter(

    x="PC1",
    y="PC2",
    by="CryptoClusters",
    hover_cols =["coin_id"],
    title = "Scatter Plot by KMeans - PCA CryptoClusters"
)

scatter_pca_clusters

:NdOverlay   [CryptoClusters]
   :Scatter   [PC1]   (PC2,coin_id)

In [241]:
# 3D Plot

df_market_pca_3d = df_market_predictions_pca_data.copy()

df_market_pca_3d["coin_id"] = df_market_data.index


fig_pca = px.scatter_3d(
    df_market_pca_3d,
    hover_name="coin_id",
    hover_data=["CryptoClusters"],
    x="PC1",
    y="PC2",
    z="PC3",
    color="CryptoClusters",
    width=800,
)

fig_pca.show()

### Visualise and Compare the Results

In this section, you will visually analyse the cluster analysis results by contrasting the outcome with and without using the optimisation techniques.

In [204]:
# Composite plot to contrast the Elbow curves

(kmeans_elbow_plot + pca_elbow_plot).cols(1)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [244]:
# Composite plot to contrast the clusters

(scatter_pca_clusters + scatter_scaled_clusters).cols(1)



:Layout
   .NdOverlay.I  :NdOverlay   [CryptoClusters]
      :Scatter   [PC1]   (PC2,coin_id)
   .NdOverlay.II :NdOverlay   [CoinCluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

In [245]:
# Composite plot to contrast the 3D clusters
fig_kmeans.show()

fig_pca.show()

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** It appears that the coin segmentation information using the DataFrame with all of the factors yielded similar results that the PCA analysis. The data is grouped in clear groups segments in both cases.